buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -272,6 +283,54 @@ private void autoConfigureClusterDependencies(
         }));
     }
 
+    private static void configureCleanupHooks(Project project) {
+        synchronized (runningClusters) {
+            if (executorService == null || executorService.isTerminated()) {
+                executorService = Executors.newSingleThreadExecutor();
+            } else {
+                throw new IllegalStateException("Trying to configure executor service twice");
+            }
+        }
+        // When the Gradle daemon is used, it will interrupt all threads when the build concludes.
+        executorService.submit(() -> {
+            while (true) {
+                try {
+                    Thread.sleep(Long.MAX_VALUE);
+                } catch (InterruptedException interrupted) {
+                    shutDownAllClusters();
+                    Thread.currentThread().interrupt();
+                    return;
+                }
+            }
+        });
+
+        project.getGradle().buildFinished(buildResult -> {
+            logger.info("Build finished");
+            shutdownExecutorService();
+        });
+        // When the Daemon is not used, or runs into issues, rely on a shutdown hook
+        // When the daemon is used, but does not work correctly and eventually dies off (e.x. due to non interruptable
+        // thread in the build) process will be stopped eventually when the daemon dies.
+        Runtime.getRuntime().addShutdownHook(new Thread(TestClustersPlugin::shutDownAllClusters));
+    }
+
+    private static void shutdownExecutorService() {
+        executorService.shutdownNow();
+        try {
+            executorService.awaitTermination(1, TimeUnit.MINUTES);
```

Comment:
Maybe log an error (or even fail more visibly) if this returns `false` and didn't actually finish terminating all the nodes?

Reply:
good catch, thanks !